experimental/CliffordAlgebras/src/CliffordAlgebras.jl

```diff
@@ -0,0 +1,589 @@
+
+Return the coefficient vector of $x$ wrt the
+canonical basis of its parent Clifford algebra,
+but all coefficients corresponding to basis elements
+with odd grading are set to zero. This also updates
+the field `x.even_coeffs`.
+"""
+function even_coefficients(x::CliffordAlgebraElem)
+  if isdefined(x, :even_coeffs)
+    return x.even_coeffs
+  end
+  _set_even_odd_coefficients!(x)
+  return x.even_coeffs
+end
+
+@doc raw"""
+    odd_coefficients(x::CliffordAlgebraElem) -> Vector
+
+Return the coefficient vector of $x$ wrt the
+canonical basis of its parent Clifford algebra,
+but all coefficients corresponding to basis elements
+with even grading are set to zero. This also updates
+the field `x.odd_coeffs`.
+"""
+function odd_coefficients(x::CliffordAlgebraElem)
+  if isdefined(x, :odd_coeffs)
+    return x.odd_coeffs
+  end
+  _set_even_odd_coefficients!(x)
+  return x.odd_coeffs
+end
+
+################################################################################
+#
+#  String I/O
+#
+################################################################################
+
+### Algebra ###
+function Base.show(io::IO, ::MIME"text/plain", C::CliffordAlgebra)
+  io = pretty(io)
+  print(io, "Clifford algebra of quadratic space with Gram matrix\n")
+  print(io, Indent())
+  show(io, "text/plain", gram_matrix(C))
+  print(io, Dedent(), "\ndefined over $(base_ring(C))")
+end
+
+Base.show(io::IO, C::CliffordAlgebra) = print(io, "Clifford algebra over $(base_ring(C))")
+
+### Elements ###
+function Base.show(io::IO, x::CliffordAlgebraElem)
+  print(io, "[")
+  foreach(y -> print(io, "$y "), coefficients(x)[1:(end - 1)])
+  print(io, "$(coefficients(x)[end])]")
+end
+
+################################################################################
+#
+#  basic functionality
+#
+################################################################################
+
+@doc raw"""
+    zero(C::CliffordAlgebra) -> CliffordAlgebraElem
+
+Return the additive identity of the Clifford algebra $C$.
+"""
+zero(C::CliffordAlgebra) = C()
+
+@doc raw"""
+    one(C::CliffordAlgebra) -> CliffordAlgebraElem
+
+Return the multiplicative identity of the Clifford algebra $C$.
+"""
+function one(C::CliffordAlgebra)
+  res = C()
+  res[1] = base_ring(C)(1)
+  return res
+end
+
+@doc raw"""
+    basis(C::CliffordAlgebra, i::Int) -> CliffordAlgebraElem
+
+Return the $i$-th canonical basis vector of the Clifford algebra $C$.
+"""
+function basis(C::CliffordAlgebra, i::Int)
+  res = C()
+  res[i] = base_ring(C)(1)
+  return res
+end
+
+@doc raw"""
+    gen(C::CliffordAlgebra, i::Int) -> CliffordAlgebraElem
+
+Return the $i$-th canonical algebra generator of the Clifford
+algebra $C$. This is just the image of the $i$-th basis vector
+of the underlying quadratic space under the canonical embedding
+into the Clifford algebra.
+"""
+function gen(C::CliffordAlgebra, i::Int)
+  res = C()
+  if i <= 0
+    res[0] #Throws a BoundsError instead of a DomainError for consistency
```

Comment:
Out of curiosity: Consistency with respect to what?

Reply:
Consistency in the sense that I think that

```julia

julia> gen(C, 0)
julia> gen(C, 3)
```
should both throw a `BoundsError`, provided the algebra `C` has only two generators